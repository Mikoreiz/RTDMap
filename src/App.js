import React, { Component } from 'react'
import { BrowserRouter as Router, Route } from 'react-router-dom'
import './App.css'
import Main from './components/Main'

export class App extends Component {
  render() {
    return (
        <Router>
          <Route exact path='/' component={Main}/>
        </Router>
    )
  }
}

export default App;