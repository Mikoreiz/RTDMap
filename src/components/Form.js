import React, { Fragment, useState, useEffect } from 'react'
import axios from 'axios'
import './styles/Form.css'

const Form = ({ handleSelect }) => {

    const [data, setData] = useState([])

    useEffect(() => {
        const getData = async () => {
            await axios.get("../json/route_list/routes.json")
            .then((res) => {setData(res.data)})
            .catch((e) => {
                console.log(e)
            })
        }
        getData()
    }, [])

    const handleSelection = (e) => {
        handleSelect(e.target.value)
    }

    return (
        <Fragment>            
            <form className="form">
                <label className='label'>Choose A Route: </label>
                <select onChange={handleSelection}>
                    { data.routes ? (
                            data.routes.map((route) => (
                                <option value={route.id} key={route.id}>{route.id}&nbsp;&nbsp;&nbsp;{route.longName}</option>
                                ))
                    )  : (
                        <option data-testid="noRoutes" value={"1"}>No routes available</option>
                    )}
                </select>
            </form>
        </Fragment>
    )
}

export default Form