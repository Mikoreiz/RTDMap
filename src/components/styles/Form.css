.form {
    display: inline-block;
    margin: 1em;
}

.label {
    font-size: 3vh;
    font-weight: 600;
    text-transform: uppercase;
    color: #800000
}

select {
    border: 1px solid #fff;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
    font-size: 2.2vh;
    font-weight: 200;
    border-bottom: 0.1em solid rgb(150, 150, 150);
}

select:active, select:hover, select:focus {
    outline: none;
}

option {
    font-size: 1vh;
    font-weight: 200;
}