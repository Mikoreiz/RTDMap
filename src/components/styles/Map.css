.routeSelection {
    width: 100%;
    height: 8.5vh;
}

.directionDiv {
    display: inline-block;
    height: 5vh;
}

.direction {
    display: inline-block;
    margin-left: 2em;
    font-size: 3vh;
    font-weight: 600;
    text-transform: uppercase;
    color: #800000
}

.reverseButton {
    position: relative;
    display: inline-block;
    top: -0.6em;
    font-family:-apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
    width: 6.5em;
    height: 3.2em;
    background-color: #800000;
    color: white;
    border: 0.2em solid rgb(245, 245, 245);
    border-radius: 0.5em;
    font-weight: 300;
    text-transform: uppercase;
    margin-left: 1em;
}

button:hover {
    background-color: #ac3737;
}

button:active {
    transform: translateY(1px);
}

.srcAndDest {
    display: inline-block;
    margin-left: 0.5em;
}

.reverseForm {
    display: inline-block;
}

.src, .dest {
    font-weight: 400;
}

.src {
    color: #EDAF00;
}

.dest {
    color: #4285F4;
}

p {
    display: inline-block;
    margin: 0;
}