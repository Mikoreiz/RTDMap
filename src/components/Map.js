// CHANGE DIRECTION BUTTON/FUNCTIONALITY AND SOURCE/DESTINATION DISPLAY

import React, { Fragment, useEffect, useRef, useState } from "react";
import L from "leaflet";

import Form from './Form'

import './styles/Map.css'
import axios from "axios";

const Map = () => {

    const mapRef = useRef(null)
    const tileRef = useRef(null)
    const layerRef = useRef(null)
    const [ chosenRoute, setChosenRoute ] = useState("1")
    const [ mapData, setMapData ] = useState({})
    const [ directionID, setDirectionID ] = useState("0")
    const [ source, setSource ] = useState("")
    const [ destination, setDestination ] = useState("")

    const changeDirection = (e) => {
        e.preventDefault()
        if (directionID === "0") {
            setDirectionID("1")
        } else {
            setDirectionID("0")
        }
    }

    // Create our map tile layer:
    tileRef.current = L.tileLayer(`https://api.mapbox.com/styles/v1/mapbox/streets-v11/tiles/{z}/{x}/{y}?access_token=pk.eyJ1IjoibWlraGFlbHJ0ZCIsImEiOiJjanMzdHdpdHAwM2Q3NDRvMW5ueWlpaXh6In0.5QRX579-FSS4GO9VRrNx1A`, {
        attribution: 'Map data &copy; <a href="https://www.openstreetmap.org/">OpenStreetMap</a> contributors, <a href="https://creativecommons.org/licenses/by-sa/2.0/">CC-BY-SA</a>, Imagery © <a href="https://www.mapbox.com/">Mapbox</a>',
        tileSize: 512,
        zoomOffset: -1
    })

    // Define the styles that are to be passed to the map instance:
    const mapStyles = {
        overflow: "hidden",
        width: "100%",
        height: "92vh",
        float: "left",
    }

    // Define an object literal with params that will be passed to the map:
    const mapParams = {
        center: [37.9545389, -121.2858201],
        layers: [tileRef.current],
        maxZoom: 19,
        maxNativeZoom: 17,
        scrollWheelZoom: false, // Prevents the map from moving when scrolling (only allows two finger scroll or mouse click)
    }
    
    useEffect(() => {
        mapRef.current = L.map("map", mapParams);
        /* eslint-disable-next-line */
    }, []);

    useEffect(() => {
        layerRef.current = L.layerGroup().addTo(mapRef.current)
    }, [])

    useEffect(() => {
        const getMapData = async () => {
            await axios.get(`../json/route_details/${chosenRoute}.json`)
            .then((res) => {setMapData(res.data.routes[0])})
            .catch((e) => {
                console.log(e)
            })
        }
        getMapData()
    }, [chosenRoute])

    useEffect(() => {
        layerRef.current.clearLayers()

        let dID = ""
        if (chosenRoute.length > 1 || chosenRoute === "5") {
            dID = directionID
        } else {
            dID = "0"
        }
        console.log(chosenRoute)

        const latlon = []

        if (mapData.shapes) {
            let max = 0
            let index = -1
            mapData.shapes.forEach(shape => {
                if (shape.locs.length > max && shape.directionId === dID) {
                        max = shape.locs.length
                }
            })

            index = mapData.shapes.findIndex((element, index) => {
                if (element.locs.length === max) {
                    return true
                }
                return false
            })
            
            mapData.shapes[index].locs.forEach(loc => {
                latlon.push([loc.lat, loc.lon])
            })

            if (chosenRoute === "2") {
                mapData.shapes[2].locs.forEach(loc => {
                    latlon.push([loc.lat, loc.lon])
                })
            }

            const polyline = L.polyline(
                latlon,
                {
                    color: "black",
                    weight: 2.5
            }).addTo(layerRef.current)

            mapRef.current.fitBounds(polyline.getBounds())
        }

        if (mapData.directions) {
            const busStops = mapData.directions[dID].stops
            setSource(busStops[0].name)
            setDestination(busStops[busStops.length - 1].name)
            console.log(source + " - " + destination)

            mapData.directions[dID].stops.forEach((stop, index) => {
                let textColor = "#800000"
                if (stop.name === source) {
                    textColor = "#EDAF00"
                } else if (stop.name === destination) {
                    textColor = "#4285F4"
                }

                L.circleMarker(
                    [stop.lat, stop.lon], 
                    { 
                        radius: 11,
                        fillOpacity: 1,
                        fillColor: textColor,
                        color: "white"
                    }).addTo(layerRef.current)
                    .bindPopup(`<h3>${index + 1} - ${stop.name}</h3>`) 
            })
        }
    }, [mapData, directionID, source, destination])

    return (
        <Fragment>
            <div data-testid="formDiv" className="routeSelection">
                <Form 
                    handleSelect={(route) => setChosenRoute(route)} 
                />
                { 
                    source !== destination ? (
                        <div className="directionDiv">
                            <label className="direction">Direction: </label>
                            <div className="srcAndDest">
                                <p><span data-testid="source" className="src">{source}</span> → <span data-testid="dest" className="dest">{destination}</span></p>
                            </div>
                            <form className="reverseForm" onSubmit={changeDirection}>
                                <button className="reverseButton" type="submit">Change Direction</button>
                            </form>
                        </div>
                    ) : (<p data-testid="loading"></p>)
                }
            </div> 
            <div data-testid="mapDiv" id="map" style={mapStyles} />
        </Fragment>
    )
}

export default Map