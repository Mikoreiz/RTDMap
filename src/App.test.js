import { render, screen, cleanup } from '@testing-library/react'
import axiosMock from 'axios'
import App from './App'
import Map from './components/Map'
import Form from './components/Form'
import userEvent from '@testing-library/user-event'

afterEach(cleanup)

test('Renders App component', () => {
  const { asFragment } = render(<App />)
  expect(asFragment(<App />)).toMatchSnapshot()
})

test('Renders Map component', () => {
  const { asFragment } = render(<Map />)
  expect(asFragment(<Map />)).toMatchSnapshot()
})

test('Renders Form component', () => {
  const { asFragment } = render(<Form />)
  expect(asFragment(<Form />)).toMatchSnapshot()
})

test('Display a loading text', () => {
  const { getByTestId } = render(<Map />)
  expect(getByTestId('loading')).toHaveTextContent('Loading...')
})

test('Display a no routes available text', () => {
  const { getByTestId } = render(<Form />)
  expect(getByTestId('noRoutes')).toHaveTextContent('No routes available')
})

