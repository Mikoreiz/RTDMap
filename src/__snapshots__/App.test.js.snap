// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders App component 1`] = `
<DocumentFragment>
  <div
    class="routeSelection"
    data-testid="formDiv"
  >
    <form
      class="form"
    >
      <label
        class="label"
      >
        Choose A Route: 
      </label>
      <select>
        <option
          data-testid="noRoutes"
          value="1"
        >
          No routes available
        </option>
      </select>
    </form>
    <label
      class="direction"
    >
      Direction: 
    </label>
    <p
      data-testid="loading"
    >
      Loading...
    </p>
    <form>
      <button
        class="reverseButton"
        type="submit"
      >
        Change Direction
      </button>
    </form>
  </div>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    data-testid="mapDiv"
    id="map"
    style="overflow: hidden; width: 100%; height: 92vh; float: left; position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Renders Form component 1`] = `
<DocumentFragment>
  <form
    class="form"
  >
    <label
      class="label"
    >
      Choose A Route: 
    </label>
    <select>
      <option
        data-testid="noRoutes"
        value="1"
      >
        No routes available
      </option>
    </select>
  </form>
</DocumentFragment>
`;

exports[`Renders Map component 1`] = `
<DocumentFragment>
  <div
    class="routeSelection"
    data-testid="formDiv"
  >
    <form
      class="form"
    >
      <label
        class="label"
      >
        Choose A Route: 
      </label>
      <select>
        <option
          data-testid="noRoutes"
          value="1"
        >
          No routes available
        </option>
      </select>
    </form>
    <label
      class="direction"
    >
      Direction: 
    </label>
    <p
      data-testid="loading"
    >
      Loading...
    </p>
    <form>
      <button
        class="reverseButton"
        type="submit"
      >
        Change Direction
      </button>
    </form>
  </div>
  <div
    class="leaflet-container leaflet-grab leaflet-touch-drag"
    data-testid="mapDiv"
    id="map"
    style="overflow: hidden; width: 100%; height: 92vh; float: left; position: relative;"
    tabindex="0"
  >
    <div
      class="leaflet-pane leaflet-map-pane"
      style="left: 0px; top: 0px;"
    >
      <div
        class="leaflet-pane leaflet-tile-pane"
      />
      <div
        class="leaflet-pane leaflet-shadow-pane"
      />
      <div
        class="leaflet-pane leaflet-overlay-pane"
      />
      <div
        class="leaflet-pane leaflet-marker-pane"
      />
      <div
        class="leaflet-pane leaflet-tooltip-pane"
      />
      <div
        class="leaflet-pane leaflet-popup-pane"
      />
    </div>
    <div
      class="leaflet-control-container"
    >
      <div
        class="leaflet-top leaflet-left"
      >
        <div
          class="leaflet-control-zoom leaflet-bar leaflet-control"
        >
          <a
            aria-label="Zoom in"
            class="leaflet-control-zoom-in"
            href="#"
            role="button"
            title="Zoom in"
          >
            +
          </a>
          <a
            aria-label="Zoom out"
            class="leaflet-control-zoom-out"
            href="#"
            role="button"
            title="Zoom out"
          >
            −
          </a>
        </div>
      </div>
      <div
        class="leaflet-top leaflet-right"
      />
      <div
        class="leaflet-bottom leaflet-left"
      />
      <div
        class="leaflet-bottom leaflet-right"
      >
        <div
          class="leaflet-control-attribution leaflet-control"
        >
          <a
            href="https://leafletjs.com"
            title="A JS library for interactive maps"
          >
            Leaflet
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
